Adding collective quantization API

marjanfariborz opened the pull request:
Summary: Created wrapper that takes the collective op and a quantization type as an arguments. It quantize the input, performs the collective op, and and perform dequantization

Test Plan: Tested through distributed_gloo_fork.

Differential Revision: D29682812

pritamdamania87 commented:
cc @raghuramank100 Can someone from the quantization side review some of the quant changes here?

pritamdamania87 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,122 @@
+def fp32_to_fp16_with_clamp(tensor: torch.Tensor) -> torch.Tensor:
+    return torch.clamp(tensor, TORCH_HALF_MIN, TORCH_HALF_MAX).half()
+
+def quntize_tensor(tensor, qtype):
```
nit: quantize

pritamdamania87 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,122 @@
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+def quntize_tensor_list(tensor_list, qtype):
```
nit: quantize

pritamdamania87 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,122 @@
+                    output_tensor_list.dtype
+                )
+            )
+        elif quant_loss is None:
+            tensor = [t.float() for t in output_tensor_list]
+            return tensor
+        elif quant_loss is not None:
+            tensor = [t.float() / quant_loss for t in output_tensor_list]
+            return tensor
```
nit: Can't we reuse `dequantize_tensor` here?

pritamdamania87 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,122 @@
+
+        if (func == dist.all_gather):
+            input_tensors = quntize_tensor(args[0], qtype)
+            out_tensors = quntize_tensor_list(tensor, qtype)
```
Why do we need to quantize the outputs? These are just empty tensors right?

pritamdamania87 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,122 @@
+
+        elif (func == dist.all_to_all):
+            input_tensors = quntize_tensor(args[0], qtype)
+            out_tensors = quntize_tensor(tensor, qtype)
```
Same as above, why do we need to quantize outputs?

pritamdamania87 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,122 @@
+        )
+
+
+def auto_quantize(func, qtype, quant_loss=None):
```
This is a public API, so we need docs for this method (See https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L1203 for an example). Also, apart from this method make rest of the methods in this file private since this is the only public API we want to expose to users.

pritamdamania87 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,122 @@
+
+
+def auto_quantize(func, qtype, quant_loss=None):
+    def wrapper(tensor, *args, **kwargs):
```
Why is the first argument of the wrapper `tensor`? For both `all_gather` and `all_to_all` the first argument is a tensor list.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2478,6 +2480,18 @@ def test_all_reduce_coalesced_sum(self):
             )
 
+        @require_backend({"gloo"})
+        def test_all_reduce_coalesced_sum_quant(self):
```
We don't have allreduce implemented yet?

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2749,6 +2763,34 @@ def _test_all_gather_helper(
 
             self._barrier()
 
+        def test_all_gather_quant(
```
Can't we just reuse `_test_all_gather_helper` here and add a `quant` parameter to `_test_all_gather_helper` which calls `auto_quantize`? That way we don't repeat a lot of the logic here.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2992,6 +3040,36 @@ def _test_all_to_all_single_unequal_split_helper(
                 self.assertEqual(out_tensor, expected_tensor)
             self._barrier()
 
+        def _test_all_to_all_helper_quantizer(
```
Isn't it simpler to just extend `_test_all_to_all_helper` instead of having another method?

pritamdamania87 reviewed (commented):
Looks like there are some unit test failures as well.

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,117 @@
+        quant = quantization_wrapper()
+        if (func == dist.all_gather):
+            input_tensors = quant.quntize_tensor(args[0], qtype)
+            out_tensors = quant.quntize_tensor_list(tensors, qtype)
```
also, making the first argument of `wrapper` with `tensors` are limiting the first argument type still? For example `dist.all_to_all_single` we have the first argument as a tensor type instead of a tensor list, so we can't for sure the first argument is always a tensor list. 

why not we just make wrapper signature be like `def wrapper(*args, **kwargs)` and extract the arguments base on the `func` type, i.e. all_gather we check if args[0] is a list, if it is we quantize_tensor_list, if it's tensor we quantize tensor.

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,117 @@
+
+        else:
+            raise RuntimeError(
+                'The collective ops {} is not supported yet'.format(func)
```
nit: can use f-string

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,117 @@
+            work = dist.all_gather(out_tensors, input_tensors, group=group, async_op=async_op)
+
+        elif (func == dist.all_to_all):
+            input_tensors = quant.untize_tensor(args[0], qtype)
```
what's "untize_tensor"?

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,117 @@
+                'The collective ops {} is not supported yet'.format(func)
+            )
+
+        for i, t in enumerate(quant.dequantize_tensor_list(out_tensors, qtype, quant_loss=quant_loss)):
```
I think this will only work if it's not an `async_op`, which means `dist.all_to_all` already finished `wait()` and `out_tensors` already filled with real content. 

What if `async_op=True`? In this case `out_tensors` is not ready, and we are dequantizing them preemptively, this is wrong as it makes the dequantization useless. What we need to do in this case is to wait until the `work` called `wait()` and `out_tensors` are ready, then apply dequantization there. 

This change might be more involved (i.e. we probably need to add `callback` functionality to `ProcessGroup::Work` and fire the callback when results is ready. This don't need to happen in this diff, we can address it later. But we need to throw error to the user in this PR to warn that the `async_op=True` mode is not supported yet.

wanchaol commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2481,6 +2483,18 @@ def test_all_reduce_coalesced_sum(self):
+        @require_backend({"gloo"})
+        def test_all_reduce_coalesced_sum_quant(self):
+            group, group_id, rank = self._init_global_test()
+            self._test_all_reduce_coalesced_helper_quant(
```
Let's remove this test as pritam suggested.

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+
+        else:
+            raise RuntimeError(
+                f"The collective ops {func} is not supported yet"
```
nit: The collective op. 

Also, can we add a test case to cover this runtime error?

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+
+def auto_quantize(func, qtype, quant_loss=None):
+    """
+    This is a prototype API that automahically quant  the input tensors, choose the precision types, and
```
automatically quantize

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+def auto_quantize(func, qtype, quant_loss=None):
+    """
+    This is a prototype API that automahically quant  the input tensors, choose the precision types, and
+    pass other necessary arguments and then dequantizes the output.
```
It's beneficial to document the argument and return types so that user of this API can understand how to use it, can you follow documentation principles in `distributed_c10d.py` to document argument and return types?

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+            "to be of type List[torch.Tensor]."
+        )
+    elif (qtype == QuantType.FP16):
+        tensor = [_dequantize_tensor(t, qtype) for t in tensors]
```
naming: dequantized_tensor_list instead of `tensor`

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+            "to be of type List[torch.Tensor]."
+        )
+    if (qtype == QuantType.FP16):
+        input_tensor = [_quantize_tensor(t, qtype) for t in tensor_list]
```
naming: quantized_tensor_list instead of `input_tensor`

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+                f"The collective ops {func} is not supported yet"
+            )
+
+        for i, t in enumerate(_dequantize_tensor_list(out_tensors, qtype, quant_loss=quant_loss)):
```
why we need to enumerate and reassign the `tensors`? We can just assign the tensor to new quantized/dequantized ones in the original list, as those lists are passed as list of references anyways.

Also can we make this specific to each call? there's no `out_tensors` in other ops like `all_to_all_single`

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+        )
+
+
+def _dequantize_tensor_list(tensors, qtype, quant_loss=None):
```
nit: also use `tensor_list` just to match with `quantize_tensor_list` signature.

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+    if (qtype == QuantType.FP16):
+        if tensor.dtype != torch.float16:
+            raise RuntimeError(
+                "tensor dtype is {} while expected to be FP16.".format(
```
nit: f-string

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+        else:
+            return tensor.float() / quant_loss
+    else:
+        raise RuntimeError(
```
ditto

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+    if not isinstance(tensor, torch.Tensor):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be a single tensor list  of type torch.Tensor."
```
what's this error message mean? should be sth like f"_dequantize_tensor expecting torch.Tensor as input but found {type(tensor)}"

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+        isinstance(p, torch.Tensor) for p in tensors
+    ):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
```
same as previously suggested.

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+        return tensor
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
```
nit: f-string

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+        return _fp32_to_fp16_with_clamp(tensor)
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
```
nit: f-string

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+        isinstance(p, torch.Tensor) for p in tensor_list
+    ):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
```
f"_quantize_tensor_list expecting torch.Tensor as input but found {type(tensor_list)}"

wanchaol reviewed (commented):
Looks much better! a few comments on doc and error message suggestions inlined.

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+import functools
+import torch
+import torch.autograd
```
it seems like we don't need the imports of `autograd` and `nn`?

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,130 @@
+
+    Currently it only supports:
+        . FP16 quantization method
+        . all_gather, all_to_all collective ops
```
can you document all the argument and return types? `func` `qtype` `quant_loss`, and return type

wanchaol reviewed (commented):
CI signals looks irrelevant to your changes. Could you rebase and see if CI errors disappear?

wanchaol commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,141 @@
+        quant_loss: (float) initialized to None.
+
+    Returns:
+        tensor_list (list[Tensor]): The output of the collective op.
```
There's no return "name", just document the return type, and explain what the return type is, the return type for the `auto_quantize` API should be a `Callable` instead. 

Please take a look at an example in c10d API https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L2412

wanchaol commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2082,16 +2085,34 @@ def _test_all_reduce_helper(
+                            tensor_shapes=tensor_shapes,
+                        )
+                        raise RuntimeError('All reduce does not support quantization')
+                    except:
```
nit: fix this lint error

jerryzh168 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,141 @@
+TORCH_HALF_MIN = torch.finfo(torch.float16).min
+TORCH_HALF_MAX = torch.finfo(torch.float16).max
+
+class QuantType(Enum):
```
we have QuantType defined here as well: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quant_type.py to reduce confusion, should we rename this to something like DQuantType (meaning distributed quant type)? or if you only want to do float16/bfloat16 I think it would be good enough to just pass around dtype directly instead of creating a new type here.

marjanfariborz replied:
The reason we are creating the new Enum class and the qtype is that we want to support other quantization methods in the future. One of my stretch goals is to add INT8 type.

jerryzh168 replied:
I think even for int8 type, we can do that with dtype probably. will you have different ways to quantize for the same dtype?

Also are we just quantizing weights? or activation as well?

wanchaol replied:
> I think even for int8 type, we can do that with dtype probably. will you have different ways to quantize for the same dtype?

yeah we can do `dtype` directly, but as you said, we do have plans to support all quantization types that `torch.quantization` offers, including BFP16 and different quantization schemes, so this type might expand to including both `dtype` and [`torch.qscheme`](https://pytorch.org/docs/stable/quantization-support.html?highlight=qscheme) to let user specify in details how to quantize the tensor before the collective call, so exposed a class here as user's API here might be good.  Do you think this is legit or have any suggestions on supporting both quantization dtype/scheme as we evolve?

> Also are we just quantizing weights? or activation as well?

I think no weights or activation concepts here, we are just trying to quantize/dequantize tensors before/after the collective, so any tensor/tensor list passed to the collective call, we will try to quantize/dequantize them?

jerryzh168 replied:
I see, there are some quantization types like static quantization that may not make sense here I feel, but we can discuss later. I think the only thing that makes sense for int8 is dynamically quantize the Tensor (that is, we calculate min/max on the single Tensor that we want to quantize and then quantize it), that's why I feel dtype is good enough. qscheme is probably a separate discussion, I think it makes sense to add qscheme in the future.

jerryzh168 commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,141 @@
+            f'Quantization type {qtype} is not supported'
+        )
+
+def _dequantize_tensor(tensor, qtype, quant_loss=None):
```
if we only support float16/bfloat16, I think dequantize would work the same way (tensor.to(torch.float)))and we probably don't need to pass around the `qtype` here?

marjanfariborz commented on torch/distributed/algorithms/quantization.py:
```diff
@@ -0,0 +1,141 @@
+
+    Args:
+        func: (str, name of the collective op) currently supporting dist.all_gather or dist.all_to_all
+        qtype: QuantType.FP16
```
Gotcha! In some of the documentations they gave an example of the input and output arguments.

wanchaol commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2800,6 +2823,11 @@ def test_all_gather_full_group(self):
             group, group_id, rank = self._init_full_group_test()
             self._test_all_gather_helper(group, group_id, rank)
 
+        @unittest.skipIf(BACKEND == "nccl", "Nccl does not support CPU tensors")
```
you probably need to import `unittest` package in this file to get rid of the CI failures :)

wanchaol commented:
https://app.circleci.com/pipelines/github/pytorch/pytorch/359042/workflows/e002ac7d-4d4b-4597-8995-975023aba06f/jobs/15146696

This seems a failure related to your change, can you take a look?

wanchaol commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3105,6 +3141,12 @@ def test_all_to_all(self):
             self._test_all_to_all_helper(group, group_id, rank)
 
+        @sandcastle_skip_if(BACKEND != "nccl", "Only NCCL supports all_to_all")
+        def test_all_to_all_quantized(self):
```
There're some test failures related to this, see https://app.circleci.com/pipelines/github/pytorch/pytorch/359751/workflows/b2601889-d460-4429-abd9-882385465852/jobs/15174372 and https://app.circleci.com/pipelines/github/pytorch/pytorch/359751/workflows/b2601889-d460-4429-abd9-882385465852/jobs/15174373

It seems like if you want to use nccl backend only, you need to pass in cuda tensors right? otherwise nccl will not accept it. You probably want to use `cuda` and `rank_to_GPU` argument if you want to use the all_to_all_helper

marjanfariborz replied:
Yeah, I'm currently adding them. Also, using GPU servers to test them internally.

wanchaol commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1,3 +1,4 @@
+import unittest
```
fix this lint, if you use `sandcastle_skip_if`, you probably won't need to import unittest

wanchaol reviewed (approved):
Looks good to me! thanks for adding this. Please rebase and make sure CI is green before landing :)

marjanfariborz committed:
Adding collective quantization API (#62142)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62142

Created wrapper that takes the collective op and a quantization type as an arguments. It quantize the input, performs the collective op, and and perform dequantization

Test Plan:
Tested through distributed_gloo_fork.
e.g., buck test mode/dev-nosan caffe2/test/distributed:distributed_nccl_fork -- test_all_to_all_quantized

Reviewed By: wanchaol

Differential Revision: D29682812

fbshipit-source-id: 6061f787d52c88eb969800560c98c38dd40f3cb8
